pyroomacoustics/signal.py

```diff
@@ -0,0 +1,73 @@
+# Signal object for simulating source signals. 
+# Copyright (C) 2019  Robin Scheibler, Ivan Dokmanic, Sidney Barthe
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+# You should have received a copy of the MIT License along with this program. If
+# not, see <https://opensource.org/licenses/MIT>.
+
+r'''
+Signal
+======
+
+Class to generate some common source signals, or read source signals from file.
+'''
+
+from __future__ import print_function
+
+import numpy as np
+
+available_types = {
+    'mono': {'frequency': 'Frequency of mono sound (Hz)', 't_max':'Maximum time (s)'},
+    'random': {'N': 'Number of random samples'}
+}
+
+class Signal(object):
```

Comment:
I have previously also thought about using a special class for signals so that we can add properties like sampling rate, etc, to it. However, I never fully proceeded due to difficulty to imagine interaction with numpy and the Signal objects in the future. Do we want the hurdle of converting from signals to numpy array, and then back ?

Regardless, at the time I looked into sub-classing the ndarray class, so that we could retain compatibility with numpy, i.e. feed a Signal to a regular numpy routine. Maybe that could be one way to do it ? Here is the numpy [doc](https://numpy.org/doc/stable/user/basics.subclassing.html). Take a look a the simple example.

Reply:
My main reason to add the class is exactly like you say, to add properties to the signal. The ndarray is simply one member of it (I call it `self.samples`), but there is also the sampling rate, the type of the signal, etc. That way we don't loose this information throughout the pipeline. Wherever the ndarray is needed, it would be enough to replace the code by `signal.samples`. 

The numpy sublcass seems like a very interesting alternative to not clog the code, especially if we want to keep compatibility with both ndarray and the Signal class. If you think this could be an option I'd be happy to implement a first example. I'd move this to a different pull request though since then it would be come a bigger things.

Reply:
Cool. In the beginning I also thought it would be neat to have extra attributes on the signals (e.g. sampling rate obviously). So if we can do it in a smart way I'm up for it.

For your specific use case, I there might also be alternatives.
1. Create a sub-package with routines pretty much like `numpy.zeros` or `numpy.ones` but for typical signals, e.g. white noise, sweep, sine, etc.
2. If the only goal is to add this signal to a sound source, we could add methods to provide the signal of the sound source. In this case, we might also need to provide a sampling frequency to the sound source to make the interface smoother.
```python
my_source = pra.SoundSource(location, fs=16000)
my_source.play_white_noise(length=1.7, power=0.5, offset=0.1)
other_source = pra.SoundSource(other_location, fs=16000)
other_source.play_sweep(length=3., type="linear", fs_low=100, fs_high=21000)
``` 

I think 1. is simplest and most straightforward, and not incompatible with adding 2. at a later time.

Reply:
Great ideas! I agree point 1 would be a nice first step, because these functions will be useful in the classes written in 2. or 3.. I have removed the Signal subclass from this pull request to keep this pull request about the Anechoic room only, but I will create a new pull request soon and include the history there.